Makefile

```diff
@@ -428,49 +441,50 @@ $(XTDLP)/$(XTENSA_TOOLCHAIN_LINUX_TAR):
 
 $(XTDLP)/$(BINUTILS_DIR)/configure.ac:
 	@echo "You cloned without --recursive, fetching $(BINUTILS_DIR) for you."
-	git submodule update --init src/$(BINUTILS_DIR)
+	git submodule update --init $(XTDLP)/$(BINUTILS_DIR)
 
 
 
 $(XTDLP)/$(NEWLIB_DIR)/configure.ac:
 #	git clone -b xtensa https://github.com/jcmvbkbc/newlib-xtensa.git $(XTDLP)/$(NEWLIB_DIR)
 	@echo "You cloned without --recursive, fetching jcmvbkbc/newlib-xtensa for you."
-	git submodule update --init src/$(NEWLIB_DIR)
+	git submodule update --init $(XTDLP)/$(NEWLIB_DIR)
 
 $(XTDLP)/$(GCC_DIR)/configure.ac:
 #	git clone https://github.com/jcmvbkbc/gcc-xtensa.git $(XTDLP)/$(GCC_DIR)
 	@echo "You cloned without --recursive, fetching jcmvbkbc/gcc-xtensa for you."
-	git submodule update --init src/$(GCC_DIR)
+	git submodule update --init $(XTDLP)/$(GCC_DIR)
 
 $(XTDLP)/$(LIBHAL_DIR)/configure.ac:
 	@echo "You cloned without --recursive, fetching submodules for you."
-	git submodule update --init src/$(LIBHAL_DIR)
+	git submodule update --init $(XTDLP)/$(LIBHAL_DIR)
 
 
 $(XTDLP)/$(ESPTOOL_DIR)/esptool.py:
 	@echo "You cloned without --recursive, fetching esptool for you."
-	git submodule update --init src/$(ESPTOOL_DIR)
+	git submodule update --init $(XTDLP)/$(ESPTOOL_DIR)
 
 $(XTDLP)/$(ESPTOOL2_DIR)/esptool2.c:
 	@echo "You cloned without --recursive, fetching esptool2 for you."
-	git submodule update --init src/$(ESPTOOL2_DIR)
+	git submodule update --init $(XTDLP)/$(ESPTOOL2_DIR)
 
 $(XTDLP)/$(MEMANALYZER_DIR)/MemAnalyzer.sln:
 	@echo "You cloned without --recursive, fetching MemAnalyzer for you."
-	git submodule update --init src/$(MEMANALYZER_DIR)
+	git submodule update --init $(XTDLP)/$(MEMANALYZER_DIR)
 
 
 
-libhal: $(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a
+build_libhal: $(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a
 
-$(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a: toolchain $(XTDLP)/$(LIBHAL_DIR)/configure.ac
-	make -C $(XTDLP)/$(LIBHAL_DIR) -f ../../Makefile _libhal
+#$(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a: toolchain $(XTDLP)/$(LIBHAL_DIR)/configure.ac
+$(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a: $(XTDLP)/$(LIBHAL_DIR)/configure.ac
```

Comment:
Hi,
I like the idea, that a make runs once, and in a second run there is 
nothing to do.
Therefore I tried different things, one is to touch the build.

For libhal a tried this and that, but no really success.

Ok, at least the following idea:

I did a more orthogonal rule with a build directory as tommie did it.
I guess separating source and build is better.
And then I adapted it to the other targets in your Makefile:

> # Libhal
> $(XTDLP)/$(LIBHAL_DIR)/build: $(XTDLP)/$(LIBHAL_DIR)/configure.ac
> 	cd $(XTDLP)/$(LIBHAL_DIR); autoreconf -i
> 	mkdir -p $(XTDLP)/$(LIBHAL_DIR)/build/
> 	cd $(XTDLP)/$(LIBHAL_DIR)/build/; PATH=$(SAFEPATH) ../$(CONF_OPT) --host=$(TARGET) --prefix=$(TOOLCHAIN)/xtensa-lx106-elf/
> 	PATH=$(SAFEPATH) $(MAKE_OPT) -C $(XTDLP)/$(LIBHAL_DIR)/build/
> 	@touch $@
>
> $(XTDLP)/$(LIBHAL_DIR): $(XTDLP)/$(LIBHAL_DIR)/build
> 	PATH=$(SAFEPATH) make $(INST_OPT) -C $(XTDLP)/$(LIBHAL_DIR)/build/
> 	@touch $@

I added build-libhal

> build-libhal: get-src build-binutils $(XTDLP)/$(LIBHAL_DIR)/build $(XTDLP)/$(LIBHAL_DIR)

and added libhal to toolchain

> $(TOOLCHAIN)/bin/xtensa-lx106-elf-gcc: $(TOOLCHAIN) $(XTDLP) ... build-second-stage-gcc build-libhal strip compress-upx

For convenience I added this test (as temporary hack)

> GCC_DIR = gcc-$(GCC_VERSION)

> ifneq "$(wildcard $(XTDLP)/gcc-xtensa)" ""
>     GCC_DIR := gcc-xtensa
> endif

so we have not to care for this difference between rtos-sdk, master, 
develop.

Please tell me your opinion about these changes, so I can make a PR.

On 08.01.2016 at 04:43 wrote Dmitry Kireev:
> In Makefile
> <https://github.com/kireevco/esp-alt-sdk/pull/20#discussion_r49156077>:
>
>>
>> -$(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a: toolchain $(XTDLP)/$(LIBHAL_DIR)/configure.ac
>> -	make -C $(XTDLP)/$(LIBHAL_DIR) -f ../../Makefile _libhal
>> +#$(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a: toolchain $(XTDLP)/$(LIBHAL_DIR)/configure.ac
>> +$(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a: $(XTDLP)/$(LIBHAL_DIR)/configure.ac
>
> What if there is no toolchain built, how will we build libhal? why do we
> remove |toolchain| dependency?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kireevco/esp-alt-sdk/pull/20/files#r49156077>.
>